File: src/data/courses/dsa/miscellaneous/meet-in-the-middle/closest-subsequence-sum/closest_subsequence_sum.py
# Time: O(2^(n/2) * n), Space: O(2^(n/2))

def minAbsDifference(nums, goal):
    n = len(nums)
    mid = n // 2
    
    def generate_sums(arr):
        sums = []
        for mask in range(1 << len(arr)):
            s = sum(arr[i] for i in range(len(arr)) if mask & (1 << i))
            sums.append(s)
        return sums
    
    left_sums = sorted(generate_sums(nums[:mid]))
    right_sums = generate_sums(nums[mid:])
    
    min_diff = abs(goal)
    
    for right_sum in right_sums:
        need = goal - right_sum
        idx = bisect_left(left_sums, need)
        
        if idx < len(left_sums):
            min_diff = min(min_diff, abs(left_sums[idx] + right_sum - goal))
        if idx > 0:
            min_diff = min(min_diff, abs(left_sums[idx - 1] + right_sum - goal))
    
    return min_diff

File: src/data/courses/dsa/miscellaneous/meet-in-the-middle/closest-subsequence-sum/code/java/ClosestSubsequenceSum.java
class Solution {
    public int minAbsDifference(int[] nums, int goal) {
        int n=nums.length,h=n/2; List<Integer> L=new ArrayList<>(),R=new ArrayList<>();
        for(int m=0;m<(1<<h);m++){int s=0;for(int i=0;i<h;i++)if((m&(1<<i))!=0)s+=nums[i];L.add(s);}
        for(int m=0;m<(1<<(n-h));m++){int s=0;for(int i=0;i<n-h;i++)if((m&(1<<i))!=0)s+=nums[h+i];R.add(s);}
        Collections.sort(R); int best=Integer.MAX_VALUE;
        for(int s:L){int need=goal-s;int idx=Collections.binarySearch(R,need);if(idx<0)idx=-idx-1;
            if(idx<R.size())best=Math.min(best,Math.abs(s+R.get(idx)-goal));
            if(idx>0)best=Math.min(best,Math.abs(s+R.get(idx-1)-goal));}
        return best;
    }
}